File: 2016/day03-2.ts
function solution(content: string) {
  const lines = content
    .split("\n")
    .map((line) => [...line.matchAll(/\d+/g)].map(Number));

  const fst = lines.map((line) => line[0]);
  const snd = lines.map((line) => line[1]);
  const trd = lines.map((line) => line[2]);
  return search(fst) + search(snd) + search(trd);
}

function search(lst: number[]) {
  return lst.filter(
    (a, index, array) =>
      index % 3 === 0 && isTriangle(a, array[index + 1], array[index + 2])
  ).length;
}

function isTriangle(a: number, b: number, c: number) {
  return a + b > c && a + c > b && b + c > a;
}

Deno.readTextFile("day03.dat")
  .then((content) => content.trim())
  .then(solution)
  .then(console.log);

File: 2016/day03_2.py
# pylint: disable=missing-module-docstring, missing-function-docstring


import re
from typing import Any


def solution(content: list[str]) -> Any:
    lines = [int(x) for line in content for x in re.findall(r"\d+", line)]
    fst = lines[::3]
    snd = lines[1::3]
    trd = lines[2::3]
    return search(fst) + search(snd) + search(trd)


def search(lst: list[int]):
    return len(
        [
            True
            for a, b, c in zip(lst[::3], lst[1::3], lst[2::3])
            if is_triangle(a, b, c)
        ]
    )


def is_triangle(a: int, b: int, c: int):
    return (a + b > c) and (a + c > b) and (b + c > a)


def main() -> None:
    with open("day03.dat", "r", encoding="utf8") as file:
        print(solution(file.readlines()))


if __name__ == "__main__":
    main()
